### PR TITLE
Evaluate functions by matching instead of unification

### PR DESCRIPTION
**Pull request by ttuegel**
Supersedes: #733, #781 
After: #800, #810

###### Reviewer checklist

- [ ] Test coverage: `stack test --coverage`
- [ ] Public API documentation: `stack haddock`
- [ ] Style conformance: `stylish-haskell`

---

### CODE REVIEW

**Comment by ttuegel** [comment]
Fixes: #769

**Comment by ttuegel** [comment]
There is some problem with Jenkins that I can't get around right now.

This is still broken anyway: it applies the `andBool` rule to get a non-matching substitution (somehow) and generates a remainder that Z3 can't handle.

**Comment by ttuegel** [comment]
> This is still broken anyway: it applies the `andBool` rule to get a non-matching substitution (somehow) and generates a remainder that Z3 can't handle.

This is not actually directly related to this pull request, and has to do with matching `#as` patterns. I will deal with that in a second pull request.

**Comment by ttuegel** [comment]
There is a test failure in `src/main/k/working/tests/collections/list-requires` due to the backend exhausting available memory.

**Comment by ttuegel** [comment]
It looks like we are expanding symbolic functions infinitely (again). I'm writing tests now to check that the relevant cases do not match.

File: kore/src/Kore/Builtin/AssociativeCommutative.hs [hunk 1]
```diff
@@ -255,100 +327,185 @@ instance TermWrapper Domain.NormalizedSet Domain.NoValue where
...
+flattenOpaque matchBuiltin (Domain.unwrapAc -> normalized) = do
+    let opaque = Domain.opaque normalized
+        (builtin, opaque') = partitionEithers (extractBuiltin <$> opaque)
+        transparent = Domain.wrapAc normalized { Domain.opaque = opaque' }
+    Foldable.foldrM concatNormalized transparent builtin
+  where
+    extractBuiltin termLike =
+        maybe (Right termLike) Left (matchBuiltin termLike)
+
+matchBuiltinMap
```
Comment (virgil-serbanuta) [inline kore/src/Kore/Builtin/AssociativeCommutative.hs]: I'd put this near matchBuiltinSet

**Pull request merged**